src/weights.jl

```diff
@@ -2,44 +2,181 @@
+function AnalyticWeights{S<:Real, V<:RealVector}(vs::V, s::S=sum(vs))
+    return AnalyticWeights{S, eltype(vs), V}(vs, s)
 end
 
 """
+    aweights(vs)
+
+Construct a `AnalyticWeights` type from a given array.
+"""
+aweights(vs::RealVector) = AnalyticWeights(vs)
+aweights(vs::RealArray) = AnalyticWeights(vec(vs))
+
+"""
     weights(vs)
 
-Construct a `WeightVec` from a given array.
+Alias for aweights(vs)
+"""
+weights(vs) = aweights(vs)
+
+@weights FrequencyWeights
+
+function FrequencyWeights{S<:Integer, V<:IntegerVector}(vs::V, s::S=sum(vs))
+    return FrequencyWeights{S, eltype(vs), V}(vs, s)
+end
+
+"""
+    fweights(vs)
+
+Construct a `FrequencyWeights` type from a given array.
+"""
+fweights(vs::IntegerVector) = FrequencyWeights(vs)
+fweights(vs::IntegerArray) = FrequencyWeights(vec(vs))
+
 """
-weights(vs::RealVector) = WeightVec(vs)
-weights(vs::RealArray) = WeightVec(vec(vs))
+    bias(w::FrequencyWeights, [corrected])
 
-eltype(wv::WeightVec) = eltype(wv.values)
-length(wv::WeightVec) = length(wv.values)
-values(wv::WeightVec) = wv.values
-sum(wv::WeightVec) = wv.sum
-isempty(wv::WeightVec) = isempty(wv.values)
+```math
+\fraction{1}{∑w - 1}
+```
+"""
+bias(w::FrequencyWeights, corrected=true) = inv(sum(w) - Int(corrected))
+
+@weights ProbabilityWeights
+
+function ProbabilityWeights{S<:Real, V<:RealVector}(vs::V, s::S=sum(vs))
+    return ProbabilityWeights{S, eltype(vs), V}(vs, s)
+end
 
-Base.getindex(wv::WeightVec, i) = getindex(wv.values, i)
-Base.size(wv::WeightVec) = size(wv.values)
+"""
+    pweights(vs)
+
+Construct a `ProbabilityWeights` type from a given array.
+"""
+pweights(vs::RealVector) = ProbabilityWeights(vs)
+pweights(vs::RealArray) = ProbabilityWeights(vec(vs))
 
+"""
+    bias(w::ProbabilityWeights, [corrected])
+
+```math
+\fraction{n}{∑w × (n - 1)}
+```
+"""
+function bias(w::ProbabilityWeights, corrected=true)
+    s = sum(w)
+
+    if corrected
+        n = length(values(w))
+        return n / (s * (n - 1))
+    else
+        return inv(s)
+    end
+end
+
+@weights ExponentialWeights
+
+function ExponentialWeights{V<:RealVector}(vs::V)
+    s = sum(vs)
+    return ExponentialWeights{typeof(s), eltype(vs), V}(vs, s)
+end
+
+"""
+    eweights(n, [λ])
+
+Constructs a `ExponentialWeights` type with a desired length `n` and smoothing factor `λ`,
+where each element is set to `λ * (1 - λ)^(1 - i)`.
+
+# Arguments
+* `n::Integer`: the desired length of the `Weights`
+* `λ::Real`: a smoothing factor or rate parameter between 0 and 1.
+    As this value approaches 0 the resulting weights will be almost equal,
+    while values closer to 1 will put higher weight on the end elements of the vector.
+"""
+function eweights(n::Integer, λ::Real=0.99)
+    n > 0 || throw(ArgumentError("cannot construct weights of length < 1"))
+    0 <= λ <= 1 || throw(ArgumentError("smoothing factor must be between 0 and 1"))
+    w0 = map(i -> λ * (1 - λ)^(1 - i), 1:n)
+    return weights(w0)
```

Comment:
Shouldn't this be `ExponentialWeights`? But why do we need a specific weights type: don't them enter into one of the above families? Or maybe they should just use a generic weights type (without any correction)?